### DISCUSSION TITLE
Cuda memory failing after RAM upgrade

### DISCUSSION THREAD

**Discussion by adamdoc** [CERN/TIGRE#612] [Q&A]
Hello, 

I am having some issues reconstructing using FBP/FDK after a RAM upgrade. I'm using CIL with Tigre backend, but this issue also seems to happen with Tigre by itself, and doesn't happen when using Astra as CIL backend

Our new config has a massive 1TB of RAM, but recon fails when the projection data is over 2GB (independent of recon volume size) with the following error
../Common/CUDA/TIGRE_common.cpp (7): Texture memory allocation fail 
../Common/CUDA/TIGRE_common.cpp (14): CBCT:CUDA:Atb out of memory
The kernel died, restarting...

However, after lowering to 768GB of RAM (or below), it seems I can recon from projection data of any size without issue (only tested to 150GB). The volume does have to be below 8GB or so which is a bit strange given I have 48GB of VRAM on the GPU but this is manageable with chunked recon.

Any advice for what may be going on appreciated, been confusing me for a couple of days now, Thanks!

Build cuda_12.5.r12.5/compiler.34177558_0
tigre                     2.6             np126py312h9bba915_0    ccpi
cil                       24.2.0             np126py312_0    ccpi

**Comment by AnderBiguri** [comment]
Heya, 
This is very weird. So, few thouhgts: the reason it happens with TIGRE, but not ASTRA, is because TIGRE uses pinned memory, i.e. uses the RAM (not just the VRAM) to do recon. this is why you can do large scale recons in TIGRE. This is used all the time, not just when the image is too big to fit, as it makes the code faster all of the time. So it makes sense that if this process is failing ("the kernel died") then it fails everytime regardless of the size. 

However, I am not sure if TIGRE has any control on this. The way the line that fails works is that it asks the operating system for memory in the RAM, and asks the operating system to "page-lock" (or "pin") the memory, i.e. to lock all the memory and force it to never be moved by other processes. From TIGRE side, this is 1 line of CUDA code with no parameters, so I am thinking that this may be more related to your system, rather than anything that I could fix from my side. Its possible that e.g. your CPU/motherboard does not support 1TB of RAM for page-locking, for example. Or that the OS version does not support it. I would dig a little bit that side, as I suspect the "out of memory" comes from TIGRE asking for the memory to the OS, and the OS saying "I can't give you that", and TIGRE/CUDA interpreting that it must be due to lack of memory. But we know its not that.  

So dig a bit more int he capabilities of your machine, I suspect that the issue is from that side. I'll check a bit more if there are issues like this from the CUDA side, but my preemptive search suggest no. 

As a side note, you say that the volume has to be below 8GB? Is this always for you?

**Reply by AnderBiguri** [reply]
@adamdoc for example, if you were using windows vista (which I really hope you are not :D) the maximum page-locked memory is 2GB. This simply limits the max amount of pinned memory. That said, note that TIGRE can be compiled without pinned memory. It will be slower, but it is permited. Not sure if CIL allows this when compiling TIGRE, but standalone TIGRE can indeed be compiled for it to not use this memory.

**Reply by adamdoc** [reply]
Hi Ander, thanks for all the information.
I've had a look and can't seem to find any issues with pinned memory limits, I'm on windows 11 with a xeon w5-3435

I can seem to allocate over the 2GB Tigre limit if I do so through pycuda and cuda.mem_alloc. This is confusing me a bit more as I agree it sounds like a PC issue rather than Tigre. I'll try to go into the Tigre source code and see if I can replicate the issue outside the Tigre environment. For now the 768GB config is enough 

And for the 8GB volume limit, this is for every RAM config I've tried, which makes sense as this is related to the VRAM on the GPU which isn't changing but obviously less than what i should be able to use. I suspect the issues may be related though as I get the same error as when the projection data is too big

**Reply by AnderBiguri** [reply]
@adamdoc couple of notes:
The pycuda function you mention allocates memory, but does not page-lock memory, which is not the same. So not necesarily contradictory.

I still don't understand the 8Gb issue, can you elaborate? TIGRE (with page locked memory, which is default) allows for larger-than-memory reconstruction. I reconstructed 64Gb volumes on a 8Gb VRAM GPU. If you are having issues with this, it could be also a memory allocation issue.

**Reply by adamdoc** [reply]
My apologies, I was using cuda.pagelocked_empty to allocate large memory array without issue

Yes, I think it's also a memory allocation issue for the volume. I get the same cuda error if either the page locked projection data in the RAM is over 2GB, or if the volume I am trying to reconstruct on the VRAM is over 8GB. The former goes away as long as I don't have the full RAM config, but the 8GB limit on the VRAM is always present. I know I should be able to recon larger volumes, and with Astra I can. It's a bit confusing as if our issue is with page-locked memory allocation, then this shouldn't happen?

**Reply by AnderBiguri** [reply]
@adamdoc ASTRA uses a more naive way to handle the memory, which is why its slower at large images. Ultimately, it doesn't use page-locked memory. I wonder if you can get help from CIL people to compile CIL-TIGRE with page-locked memory turned off, which is just a standard flag when compiling TIGRE. Maybe that will help verify if this is the issue.

**Reply by adamdoc** [reply]
Okay we can ignore the 8GB limit, the issue only exists with FBP. I'll be using FDK and just tested on that and I can recon a 60GB dataset no issue

I'm speaking to the CIL people now on discord, thanks for all the advice and I'll report back if I find the issue!

**Reply by AnderBiguri** [reply]
@adamdoc ah, the parallel beam projectors are not multi-GPU, as you can reconstruct slice-per-slice.

**Reply by adamdoc** [reply]
Hi again @AnderBiguri ,

I finally got round to trying a few more things and it does seem to be the page-locked memory issue

I've been able to replicate the issue using Tigre in Matlab ("out of memory" error). And if I compile using compile('-no_pinned_mem') it works ! 

I also tried again with pycuda and although it doesn't give an error when calling cuda.pagelocked_empty(array_size), the screen freezes for a few seconds so something isn't right. This only happens for array_size >2GB

Am I likely to see a significant performance decrease using non-pinned memory in Matlab? Overall I'd still like to get this working, and as it seems to be a cuda issue I'll raise something on the forums over there

**Reply by AnderBiguri** [reply]
@adamdoc no, it may be a tiny bit slower, but nothing extremely significant, you should be fine! You should be able to Compile the python one with that flag too, if you prefer to work in python.

**Reply by adamdoc** [reply]
@AnderBiguri 
I'm having a bit of trouble with this, I'm still learning python. Can I check if the following should work?

within a conda env
`git clone https://github.com/CERN/TIGRE.git`
`pip install -r requirements.txt`
`python setup.py install --no_pinned_memory`

I'm getting the following error
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\ReconMXCN\anaconda3\envs\tigre\lib\site-packages\pytigre-3.0.0-py3.9-win-amd64.egg\tigre\__init__.py", line 24, in <module>
    from .utilities.Ax import Ax
  File "C:\Users\ReconMXCN\anaconda3\envs\tigre\lib\site-packages\pytigre-3.0.0-py3.9-win-amd64.egg\tigre\utilities\Ax.py", line 4, in <module>
    from _Ax import _Ax_ext
ImportError: DLL load failed while importing _Ax: The specified module could not be found.

**Reply by AnderBiguri** [reply]
you need to install with `pip install .` unfortunately, which indeed may mean that the --no_pinned_memory flag may not be working. You can however go inside `setup.py` and change the default value of it, so it compiles without it by default.

**Reply by adamdoc** [reply]
@AnderBiguri I'm still getting the same error with 
`pip install .`

I forgot to mention I get this error when I run 
`import tigre`
I see pytigre 3.0.0 in my package list, so it's installed just not correctly

**Reply by AnderBiguri** [reply]
maybe the `python setup.py install` has configured something badly. I suggest restarting from scratch the isntallation. The import error is weird, as `pip install .` should install, but not import.

**Reply by adamdoc** [reply]
Apologies @AnderBiguri I didn't explain very well. The error I was seeing was after installing and running `import tigre` in python. But I found the solution, there is a fix in a `__init__.py` for missing dll - so all working now

After many restarts I'm glad to say I was able to build TIGRE with `no_pinned=True` in `setup.py` my CIL environment, and have just reconstructed a big dataset without issue. Thanks for all the help!!

**Reply by AnderBiguri** [reply]
@adamdoc what is the fix? this may be a thing I need to fix in TIGRE

**Reply by adamdoc** [reply]
It's the issue with Windows and Python 3.8
https://github.com/CERN/TIGRE/issues/349

Which is fixed when uncommenting lines 9-19 here 
https://github.com/CERN/TIGRE/blob/master/Python/tigre/__init__.py
